### PR TITLE
Add support for `Enum`

### PR DESCRIPTION
**Pull request by vreuter**
Hi @alimanfoo , I wanted to use this great library to document enums but couldn't get it to work without some addition. I wanted to get a working implementation, and I propose this as a starting point. I gave it a first pass with some of the following consideration...

I went with the implementation here to try for minimal repetition even with some shared functionality, while also trying to minimise the impact on the existing implementation of the core `_doc` function. I also gave the version for enum much more limited capability (separate internal function with pared-down signature) since some of the fields don't really make sense for `Enum`. I also left off things of questionable / arguable relevance like "notes" and "references", thinking that in general these are likely to be useful in relatively rare circumstances, when an `Enum` is being used to model some cases for a complex use case / domain / context. The downside is that this then falls under a separate decorator. I made peace w/ this thinking that any refactor which changed something to/from an enum would be significant enough to make the necessary change to the decorator be comparatively minor in comparison.

WDYT?

### CODE REVIEW

**Comment by alimanfoo** [comment]
Hi @vreuter, very happy for numpydoc to support this if it would be useful. I generally try to stick as close as possible to features and conventions defined in the numpydoc style guide, and I wonder if here we could consider documenting enums as a special case of documenting classes? The numpydoc style guide has a [section on documenting classes](https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes). If we followed that, then the members of an enum could be documented as "attributes" rather than "parameters"?

In other words, maybe we could think about the more general case of supporting documentation of classes, and then hopefully that would also cover enums too?

**Comment by vreuter** [comment]
> In other words, maybe we could think about the more general case of supporting documentation of classes, and then hopefully that would also cover enums too?

💯 good idea, I will try updating with more general support, then tailored (if needed) with slight modification to support `Enum`

**Comment by alimanfoo** [comment]
Cool. 

The numpydoc guide says the "parameters" section should refer to the constructor, so I guess we should look at the signature of `__init__` to build that section.

I guess also the "attributes" section would always have to be built from what is provided to the `@doc` decorator because there's no easy or guaranteed way to automatically populate that section.

**Comment by alimanfoo** [comment]
Now thinking there's probably a decent amount of work in getting full support for classes working, so would be totally reasonable to start with supporting enums only in this PR if that would make life easier.

Also thinking there's probably a few specialisations that we'd want for enums that wouldn't apply for classes, e.g., showing the values of the enum members. So code paths for enums and general classes might not have that much in common.

If we were to just implement support for enums for now, thinking through what else I'd suggest for this PR, I think it would be:

(1) Make `@doc()` a common decorator for handling either functions or enums, so users don't have to think about which decorator to apply (but could split into different functions internally based on type of argument, e.g., `_doc_function` and `_doc_enum`); 

(2) Change "parameters" to "attributes" for documenting members of the enum; 

(3) Inspect the contents of `__members__` and automatically add minimal documentation for any members not provided in `@doc(attributes={...})`, as well as raise error if any members given in `@doc(attributes={...})` are not actually present in the enum. 

(4) Include the value of the members in the docstring.

So, e.g., for the test currently included, would expect:

```
This tests @doc for an enum.

Attributes
----------
First = "1st"
    This is the 1st value.
```

**Comment by vreuter** [comment]
OK cool makes sense, I like all these ideas @alimanfoo and will update the PR accordingly. Probably not before this weekend, though 😅

**Comment by alimanfoo** [comment]
Cool, no rush at all, enjoying exploring these ideas!

**Comment by vreuter** [comment]
@alimanfoo I updated the implementation to comply with the four-point spec you outlined above. The diff expands a lot with the nesting of what was previously the only code path through `_doc` now under one of two branches of a conditional (on whether the given callable is an enum or not), but that's largely about indentation. Mainly, I just updated to add some test cases of what you'd spec'd out, and to have the two code paths under the same decorator for better user-side experience. The current implementation does _not_ preclude someone from passing potentially, enum-irrelevant keyword arguments to  `@doc` when documenting an enum, though. Checks could be added for that to enforce the difference, but I'm not sure how much value is to going down that route, as it represents an erroneous use case that seems most likely to arise from a refactor where someone changes what was once another callable to an enum.

WDYT?

**Comment by alimanfoo** [comment]
Cool, thanks @vreuter. I'm on leave this week but will take a look as soon as I'm back.
